WIP: Ratcliff Diffusion Model pdf and simulators

kiante-fernandez opened the pull request:
I have added a PDF file containing the Ratcliff Diffusion Model and two methods for simulating the diffusion process. However, the PDF code still needs testing as there are some remaining bugs. I wanted to share the progress so that we can identify the tests that need to be conducted and address any issues.

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -262,4 +262,187 @@ logpdf(d::RatcliffDDM, data::Tuple) = logpdf(d, data...)
+    D = σ^2 / 2
+
+    # Program specifications
+    ϵ = 2.220446049250313e-16  # precision from 1.0 to next double-precision number
```
You can use `eps()` here instead.

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -262,4 +262,187 @@ logpdf(d::RatcliffDDM, data::Tuple) = logpdf(d, data...)
+        η = 3
+    end
+
+    if η == 0
```
I like the idea of suggesting typical parameter ranges in the documentation, but I don't think we should enforce bounds unless they are strict (e.g., a standard deviation must be non-negative). I think the best approach is to allow users to enforce their own parameter bounds. This will also be consistent with the approach taken for the other models in the package. .

kiante-fernandez replied:
Okay. I will make a separate PR that contains an implementation for this across models.

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -262,4 +262,187 @@ logpdf(d::RatcliffDDM, data::Tuple) = logpdf(d, data...)
+
+    # Initialize output vectors
+    rts = zeros(N)
+    choice = zeros(N)
```
To be consistent with the other models, I think `choice `should be a vector of integers, e.g., `zeros(Int, N)` or `fill(0, N)`

itsdfish commented:
@kiante-fernandez, thanks for the progress update. I realize this is a WIP, but I left a few comments that may still be applicable. 

My approach to tests is to start by plotting the pdf over the histogram of simulated data. That is part of the reason I include those plots in the documentation. I also like to compare the pdf to a kernel density of the pdf. It's not perfect, but it gives you some idea of whether the generative model and the pdf are consistent with each other. I also test against established libraries when available. Another thing I do is develop a test whenever I find a bug to ensure there are no regressions in the future.

DominiqueMakowski commented:
Could we also add (in the docstrings?) a brief mention about the difference between DDM and RatcliffDDM and maybe a cross-link ("See also `DDM()`")

itsdfish commented:
> Could we also add (in the docstrings?) a brief mention about the difference between DDM and RatcliffDDM and maybe a cross-link ("See also `DDM()`")

If we can get this model working, another option would be to integrate everything into one model given that `DDM` is currently a special case in which the cross-trial variability is zero. The integration procedures can be skipped in the special case. We could avoid breaking changes by using the type `DDM` for the general model.

kiante-fernandez commented:
@itsdfish I have added some updated PDF along with the CDF code, but I am getting a precompile error that might be related to some bugs with Bijectors.

itsdfish commented:
Thanks @kiante-fernandez for your hard work. It is difficult for me to look at a fork locally. Based on the unit [tests](https://github.com/itsdfish/SequentialSamplingModels.jl/actions/runs/5444977664/jobs/9903568899?pr=29#step:6:281), it seems like there is a variable called x that is not defined somewhere. It's not clear to me where its at based on the error trace. I recommend using Revise.jl and commenting out files or parts of files until you can locate the issue. 

I'll leave a few comments for now and look through the code in more detail probably tomorrow.

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -1,371 +1,662 @@
+#  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#  * General Public License for more details.
+#  *
+#  * You should have received a copy of the GNU General Public License
```
As I mentioned during our initial conversation, the GNU license is a problem because it is viral unlike the MIT licence. So if we add the GNU license to this repo, it will become GNU, and so will any code that uses this package.  I recommend reading through any papers that their code is based on and using that as a guide. Basically, it shouldn't look like you did a verbatim translation of their code into Julia. In some cases, there might be some degree of similarity because there is only one way to implement certain ideas. Does that make sense?

The other option would be to do a direct translation with the same naming conventions and put the model in its own package which could extend the same interface defined here.

DominiqueMakowski replied:
> The other option would be to do a direct translation with the same naming conventions and put the model in its own package which could extend the same interface defined here.

I think paraphrasing the code to avoid copyright issues and keeping everything in one package would be the best for users

itsdfish replied:
Thanks! Sorry. I realize it's a pain. There are somethings about  GNU (GPL) that I like better than MIT, but the main drawback with GNU that many people will not use your code if it is under GNU.  I can't say I fully understand the minutia, but it seems like it could cause issues.

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -1,371 +1,662 @@
+#
+################################################################################################################################################################
+
+TUNE_PDE_DT_MIN = 1e-6
```
In 99.9% of cases, global variables are not a good idea in any language, but they are particularly bad for performance in Julia. I think there are two options:

1. If you don't think these variables will be changed, you can define them in the functions where they are used. It looks like they are only used in one function each. So this might be a viable solution.

2. If you think people will change the values, you could define a struct containing the constants and add it to the model struct. The constants could be initialized it in the model constructor based on the keyword arguments with default values .

itsdfish commented on .vscode/settings.json:
Can you add ./vscode to the .gitignore file and remove the folder from the cache? This source will help with the later: https://stackoverflow.com/questions/1274057/how-do-i-make-git-forget-about-a-file-that-was-tracked-but-is-now-in-gitignore

DominiqueMakowski commented:
How is it going @kiante-fernandez I hope this didn't give you a nervous breakdown 😅 Let us know if you need help or support :)

kiante-fernandez commented:
> How is it going @kiante-fernandez I hope this didn't give you a nervous breakdown 😅 Let us know if you need help or support :)
No worries thanks for checking in. I just need to write some checks for the PDF and CDF. Otherwise the methods are all written. I am at a conference this weekend. Will take a look Sunday.

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -1,371 +1,605 @@
-
-# """
-# function _simpson_1D(x::Real, c::Real, ν::Real, η::Real, α::Real, z::Real, τ::Real, ϵ::::Real, lb_z::Real, ub_z::Real, n_sz::Int, lb_t::Real, ub_t::Real, n_st::Int)
+function _pdf_sv(d::RatcliffDDM{T}, rt::Real; ϵ::Real = 1.0e-12) where {T<:Real}
```
I think the parametric type information can be removed  here. Specifying parametric types is useful in two cases: dispatching based on the type (here it will always be a subtype of `Real`), and (2) using the type `T` to initialize the type of other variables. Since the function does not do any of those, I think it would be good to simplify the method signature.

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -1,371 +1,605 @@
+function _pdf_sv(d::RatcliffDDM{T}, rt::Real; ϵ::Real = 1.0e-12) where {T<:Real}
+    (ν, α, τ, z, η, sz, st, σ) = params(d)
+
+    if η == 0
```
I think the module name can be removed since there are no naming conflicts.

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -1,371 +1,605 @@
+end
+
+#use numerical integration for variability in non-decision time and bias (Ratcliff and Tuerlinckx, 2002)
+function _pdf(d::RatcliffDDM{T}, rt; ϵ::Real = 1.0e-12, n_st::Int=2, n_sz::Int=2)  where {T<:Real}
```
For these internal functions, can you add docstrings that briefly explain their role and the meaning of the inputs?  This will help anyone who needs to debug or understand the code.

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -1,371 +1,605 @@
+
+logpdf(d::RatcliffDDM, choice, rt; ϵ::Real = 1.0e-12) = log(pdf(d, choice, rt; ϵ))
+
+function logpdf(d::RatcliffDDM, data::T) where {T<:NamedTuple}
```
I think this function is missing `ϵ::Real = 1.0e-12`

itsdfish replied:
Actually, I think you can delete this method if you modify this generic [method](https://github.com/itsdfish/SequentialSamplingModels.jl/blob/7f9151d504ef6a1163154927d5515da8a9ff684c/src/utilities.jl#L51) to be 

`logpdf(d::SSM2D, data::NamedTuple; kwargs...) = logpdf.(d, data.choice, data.rt;  kwargs...)`

When the method above is called, it will pass the inputs to `logpdf(d::RatcliffDDM, choice, rt; ϵ::Real = 1.0e-12)` where the default value of `ϵ` will be used if not keyword arguments are passed.

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -1,371 +1,605 @@
+    return sum(logpdf.(d, data...))
+end
+
+function logpdf(dist::RatcliffDDM, data::Array{<:Tuple,1})
```
I recently deleted this from the LBA code. Can you delete this function?

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -1,371 +1,605 @@
+    return LL
+end
+
+logpdf(d::RatcliffDDM, data::Tuple) = logpdf(d, data...)
```
Same here. This method can be deleted.

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -1,371 +1,605 @@
+function rand(rng::AbstractRNG, d::RatcliffDDM)
+    # method::Char = "rejection"
+    return _rand_rejection(rng, d)
+    # method::Char = "stochastic"
```
Once you are confident you do not need commented code, can you delete throughout?

itsdfish commented on src/RatcliffDDM.jl:
```diff
@@ -1,371 +1,605 @@
+    #     return pdf(SequentialSamplingModels.DDM(ν, α, τ, z), t; ϵ)  + (  ( (α*z*η)^2 - 2*ν*α*z - (ν^2)*t ) / (2*(η^2)*t+2)  ) - log(sqrt((η^2)*t+1)) + ν*α*z + (ν^2)*t*0.5
+    # end
+    # return pdf(SequentialSamplingModels.DDM(ν, α, τ, z), t; ϵ)  + (  ( (α*(1-z)*η)^2 + 2*ν*α*(1-z) - (ν^2)*t ) / (2*(η^2)*t+2)  ) - log(sqrt((η^2)*t+1)) - ν*α*(1-z) + (ν^2)*t*0.5
+    return _pdf(SequentialSamplingModels.DDM(ν, α, τ, z), rt; ϵ)  + (  ( (α*z*η)^2 - 2*ν*α*z - (ν^2)*(rt-τ) ) / (2*(η^2)*(rt-τ)+2)  ) - log(sqrt((η^2)*(rt-τ)+1)) + ν*α*z + (ν^2)*(rt-τ)*0.5
```
A few weeks ago I mentioned the idea of integrating the code into a single model. The logic is that DDM is a special case of RatcliffDDM. It should be possible to skip integration for parameters with no variability. From what I can tell, the code is nearly set up to do so. Would you be able to integrate `_pdf(DDM)` into this code? If you do that, DDM.jl can be deleted and the name RatcliffDDM can be simplified to DDM.

itsdfish commented:
@kiante-fernandez, I just saw a banner at the top of the this page indicating you are requesting a review. I didn't receive a notification. Sorry if you have been waiting for a while. I should have some time on Friday to review. In the meantime, I noticed that some unit tests do [not pass](https://github.com/itsdfish/SequentialSamplingModels.jl/actions/runs/5608614465/job/15337162244?pr=29).

itsdfish commented on src/SequentialSamplingModels.jl:
You can remove this export. I simplified the SSM interface.

itsdfish commented on src/utilities.jl:
```diff
@@ -69,8 +69,8 @@ Computes the likelihood for a 2D sequential sampling model.
-logpdf(d::SSM2D, data::NamedTuple) = logpdf.(d, data.choice, data.rt)
-logpdf(d::SSM2D, data::Vector{Real}) = logpdf(d, Int(data[1]), data[2]) 
+logpdf(d::SSM2D, data::NamedTuple; kwargs...) = logpdf.(d, data.choice, data.rt; kwargs...)
+logpdf(d::SSM2D, data::Vector{Real}; kwargs...) = logpdf(d, Int(data[1]), data[2]; kwargs...) 
```
`loglikelihood` and `pdf` will need methods for `data::{Real}`

itsdfish replied:
Please disregard. That was something I missed. I added it in the latest commit.

itsdfish commented on docs/src/DDM.md:
```diff
@@ -27,8 +31,11 @@ In the code below, we will define parameters for the DDM and create a model obje
 
 The average slope of the information accumulation process. The drift gives information about the speed and direction of the accumulation of information. Typical range: -5 < ν < 5
 
+Across-trial-variability of drift rate. Standard deviation of a normal distribution with mean v describing the distribution of actual drift rates from specific trials. Values different from 0 can predict slow errors. Typical range: 0 < η < 2. Default is 0.
```
In order for the code to run correctly across `@example` blocks, you need to add an `@setup` block to define global variables. For example, in the LBA docs, we have:

```
```@setup lba
using SequentialSamplingModels
using SSMPlots 
using Random

ν=[2.75,1.75]
A = .8
k = .5
τ = .3

dist = LBA(;ν, A, k, τ) 
choices,rts = rand(dist, 100)
plot(dist)
histogram!(dist)
```
```
I don't think all of the variables defined there are necessary. So you could reduce the setup to only the necessary global variables if you would like. 

Please add the `@setup` block and test the docs locally with the following procedure:

1. in shell mode, cd to .julia/dev/SequentialSamplingModels
2. in package mode run `activate ./docs`
3. In REPL mode, run `using LiveServer; servedocs()`
4. Using `ctrl` + click, on the hyperlink `http://localhost:8000/ ...` which can be found at the end of the output

This will allow you to verify that the docs render correctly.

DominiqueMakowski commented:
How's it going @kiante-fernandez with this? If you don't have bandwidth currently, could we lay out the to-do list required to complete this in a more explicit way and maybe I can try to find some help (perhaps GPT4 can be useful ^^)

kiante-fernandez commented:
I did not see that @itsdfish got back to me. I would like to take some time to review updates to the code base first so I can make sure the implementation is in line with what we have.

itsdfish commented:
@kiante-fernandez, that sounds good. If I recall correctly, there were three  outstanding issues: (1) fix the NaNs in the pdf and logpdf, (2) fix the unit tests, and (3) make sure the docs render correctly. Let me know if you have any questions.

itsdfish commented:
@kiante-fernandez, I hope you are doing well. I wanted to follow up with you about the status of the PR and a related question about `rand`. I was hoping to use `rand`, but I am having trouble reproducing results with RT dists. I thought maybe I am doing something incorrectly, or I am misunderstanding the parameterization of the model. Here is what I have:


**Julia**
```
using SequentialSamplingModels
using SequentialSamplingModels: RatcliffDDM

dist = RatcliffDDM(ν = 1.0,α = 0.80,τ = 0.30,z = 0.25,η = 0.16,sz = 0.05, st = .10,σ = 1.0)
choice,rt = rand(dist, 10_000)

mean1 = mean(rt[choice .== 1])
prob1 = mean(choice .== 1)

mean2 = mean(rt[choice .== 2])
prob2 = mean(choice .== 2)

[mean1, mean2, prob1, prob2]
```

```
4-element Vector{Float64}:
 0.49065225546420305
 0.3867084522660908
 0.4116
 0.5884
```
**R**

```
library("rtdists")

# z = a * .25
sim_data = rdiffusion(10000, a=.8, v=1, t0=.30, z = .20, sz = .05, sv = .16,
           st0 = .10, s = 1)

data_lower = sim_data$rt[sim_data$response == "lower"]
mean_lower = mean(data_lower)
prob_lower = mean(sim_data$response == "lower")

data_upper = sim_data$rt[sim_data$response == "upper"]
mean_upper = mean(data_upper)
prob_upper = mean(sim_data$response == "upper")
c(mean_lower,mean_upper, prob_upper, prob_lower)
```
`0.4380115 0.5396267 0.4166000 0.5834000
`

The response probabilities are similar, but the ordering of the rts is reversed and Julia is a bit smaller. Any ideas?

kiante-fernandez commented:
Thanks for following up. With the holiday here, I should have some time to take a look. Also, from the example you gave, you have set z to two different values (Julia: z = 0.25 and R: z = 0.20).

itsdfish commented:
Thanks for taking a look.  

You are right that the values for `z` are different. The documentation for rtdist was a bit unclear. However, the examples lead me to believe it is absolute, so the appropriate value is z = .80 * .25 = .20. I did try the absolute value, but the results diverged even more.  I created a random walk version because it is easier to understand. Interestingly, the results between that and your code are similar. I will share that code in case it is helpful for determining a ground truth. 

Anyways, have a good break! 

```
function rand(dist::DDM; Δt=.001)
    (;ν,α,z,σ,τ,η,st,sz) = dist
    t = 0.0
    z′ = α * z 
    Δx = σ * √(Δt)
    x = sz == 0 ? z′ : rand(Uniform(z′ - sz / 2, z′ + sz / 2))
    ν′ = rand(Normal(ν, η))
    p = .50 * (1 + (ν′ * √Δt) / σ)
    while (x < α) && (x > 0)
        x += rand() ≤ p ? Δx : -Δx
        t += Δt
    end
    choice = (x < α) + 1
    t += st == 0 ?  τ : rand(Uniform(τ - st / 2, τ + st / 2))
    return choice,t
end

function rand(dist::DDM, n::Int; Δt=.001)
    choices = fill(0, n)
    rts = fill(0.0, n)
    for i ∈ 1:n 
        choices[i],rts[i] = rand(dist; Δt)
    end
    return (;choices,rts)
end 
```

kiante-fernandez commented:
@itsdfish That is my understanding as well. I tried to hunt down the sampling function `rtdist` uses. But I also confirmed this using a few different simulation functions across both R and Python.

```{r}

# Load necessary libraries
library(ggplot2)
library(rtdists)
library(Rhddmjags)

# Function to simulate and get mean RT for a given st0 value
simulate_mean_rt <- function(st0, sample_method = 1) {
  if (sample_method == 1) {
    sim_data <- rdiffusion(100000, a = 0.8, v = 1, t0 = 0.30, z = 0.20, 
                           sz = 0.0, sv = 0.0, st0 = st0, s = 1)
    mean_rt <- mean(sim_data$rt)
  } else if (sample_method == 2) {
    sim_data <- Rhddmjags::simulratcliff(100000, Alpha = 0.8, Nu = 1, Tau = 0.30, Beta = 0.20, 
                                         rangeBeta = 0.0, Eta = 0.0, rangeTau = st0, Varsigma = 1)
    mean_rt <- mean(abs(sim_data))
  } else {
    sim_data <- Rhddmjags::simul_ratcliff_slow(100000, Alpha = 0.8, Nu = 1, Tau = 0.30, Beta = 0.20, 
                                               rangeBeta = 0.0, Eta = 0.0, rangeTau = st0, Varsigma = 1, nsteps = 1000)
    mean_rt <- mean(abs(sim_data))
  }

  return(mean_rt)
}

st0_values <- seq(0, 0.7, by = 0.05)
# Define a range of st0 values and store results
results1 <- data.frame(st0 = st0_values, mean_rt = sapply(st0_values, simulate_mean_rt))
results2 <- data.frame(st0 = st0_values, mean_rt = sapply(st0_values, simulate_mean_rt, sample_method = 2))
results3 <- data.frame(st0 = st0_values, mean_rt = sapply(st0_values, simulate_mean_rt, sample_method = 3))

results1$method <- "fast_dm"
results2$method <- "rejection_rhddmjags"
results3$method <- "euler_rhddmjags"

results <- rbind(results1, results2, results3)

# Plot using ggplot2
ggplot(results, aes(x = st0, y = mean_rt, group = method, color = method)) +
  geom_line() +
  geom_point() +
  labs(title = "Mean Response Time by st0", 
       x = "st0 Value", 
       y = "Mean RT") +
  theme_minimal()
```
![sim_res](https://github.com/itsdfish/SequentialSamplingModels.jl/assets/61021880/64015566-ec03-43c7-a333-d62ccc84ca39)

Here fast_dm is the `rdiffusion` function

```{python}
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import hssm
from ssms.basic_simulators.simulator import simulator

v_true, a_true, z_true, t_true, sz_true, sv_true, st_true, s  = [1, 0.8, 0.2, 0.3, 0, 0, 0, 1]

# Define the range of st_true values
st_true_values = np.linspace(0, 1, 10)

# Initialize lists to store the mean response times and st_true values
mean_rts = []
st_values = []

# Loop over the range of st_true values
for st_true in st_true_values:
    # Simulate data
    sim_out = simulator([v_true, a_true, z_true, t_true, sz_true, sv_true, st_true, s], model="full_ddm", n_samples=10000)
    
    # Turn data into a pandas dataframe
    dataset_plots = pd.DataFrame(
        np.column_stack([sim_out["rts"][:, 0], sim_out["choices"][:, 0]]),
        columns=["rt", "response"],
    )
    
    # Round the response times to 4 decimal places
    dataset_plots['rt'] = dataset_plots['rt'].round(4)
    
    # Calculate the mean response time and store it
    mean_rt = dataset_plots['rt'].mean()
    mean_rts.append(mean_rt)
    
    # Store the st_true value
    st_values.append(st_true)

# Plot the mean response times by st_true values
plt.plot(st_values, mean_rts)
plt.xlabel('st_true')
plt.ylabel('Mean response time')

# Make the y-axis larger
plt.ylim(0, max(mean_rts) * 1.1)

plt.show()
```

So at first glance, we might have found something wrong with `rdiffusion`? 

Also, just so I am clear on the PR, I am still writing DDM.jl to replace RatcliffDDM.jl. So, when we have non-zero values of the across-trial parameters, it just calls those functions (the RatcliffDDM) instead?

itsdfish commented:
@kiante-fernandez, thanks for looking through the issue above. It does appear to be a bug. I will submit an issue later today or tomorrow. I think the code above can be repurposed for unit tests at some point. It would be good to include a few tests to make sure changes in parameters have the expected effect (or lack of effect). It goes to show that it is easy to make a simple error. I would be happy to add those tests if it would help. 

Yeah. I think it would be good to remove the DDM code and use the RatcliffDDM code because it subsumes the DDM code. There are probably different ways to approach it. Perhaps the simplest way is to use conditional checks to skip calculations needed for st, eta and sz if the value is equal to zero. For example, if you perform an expensive operation (e.g., a convolution) in the RatcliffDDM code, we can potentially skip it if eta = 0. Of course,  you will probably have a better idea of what approach is the best in terms readability, efficiency, and maintainability.  But I think this will be simpler overall because DDM is a special case where st, eta,sz = 0 in RatcliffDDM. Once all of that is done, I recommend switching the name from RatcliffDDM to DDM, which will prevent a breaking change (e.g., we are just making DDM more general).  Let me know if you have any questions or concerns.

kiante-fernandez commented:
@itsdfish I have finally had a chance to take a look through the new codebase; great work on organizing things! Following the conventions, I added the AbstractDDM and moved everything we needed to DDM.jl. I am still failing some of the tests I set up, but I thought I would ping for feedback while I have time to have eyes on this for a few days.

itsdfish commented on .vscode/settings.json:
In addition to adding `/.vscode` to git ignore, you will have to delete this file and follow the instructions [here](https://stackoverflow.com/questions/1274057/how-do-i-make-git-forget-about-a-file-that-was-tracked-but-is-now-in-gitignore) to fully remove it.  Unfortunately, removing files is somewhat of a pain if you forget to add them to .gitignore

itsdfish commented on src/DDM.jl:
```diff
@@ -1,81 +1,182 @@
-- `α`: boundary threshold separation. The amount of information that is considered for a decision. Typical range: 0.5 < α < 2
-- `τ`: non-decision time. The duration for a non-decisional processes (encoding and response execution). Typical range: 0.1 < τ < 0.5 
-- `z`: starting point. Indicator of an an initial bias towards a decision. The z parameter is relative to a (i.e. it ranges from 0 to 1).
+    - `ν`: drift rate. Average slope of the information accumulation process. The drift gives information about the speed and direction of the accumulation of information. Typical range: -5 < ν < 5
```
The parameter descriptions need to be aligned with the # Parameters header. Otherwise, they will not render.

itsdfish commented on src/DDM.jl:
```diff
@@ -125,230 +226,493 @@ function _large_time_pdf(u::T, z::T, K::Int) where {T<:Real}
+logpdf(d::DDM, choice, rt; ϵ::Real = 1.0e-12) = log(pdf(d, choice, rt; ϵ))
+
+
+"""
```
Can you please comment out the doc strings for cdf? I think it will show up in the documentation even though there is not a method for DDM.

itsdfish commented on src/DDM.jl:
```diff
@@ -125,230 +226,493 @@ function _large_time_pdf(u::T, z::T, K::Int) where {T<:Real}
+    
+method simulating the diffusion process: 
+_rand_rejection uses Tuerlinckx et al., 2001 rejection-based method for the general wiener process
+_rand_stochastic uses the stochastic Euler method to simulate the stochastic differential equation
```
What is the status with these different  methods for `rand`? If you are going to implement them later, can you remove the doc string  comment for `_rand_stochastic` for the time being?

itsdfish commented on src/DDM.jl:
```diff
@@ -125,230 +226,493 @@ function _large_time_pdf(u::T, z::T, K::Int) where {T<:Real}
+
+    # Program specifications
+    ϵ = eps()  # precision from 1.0 to next double-precision number
+    Δ = ϵ
```
Since delta = episilon, can you delete delta and use delta throughout?

itsdfish commented on src/DDM.jl:
```diff
@@ -125,230 +226,493 @@ function _large_time_pdf(u::T, z::T, K::Int) where {T<:Real}
+        μ = ν + r1 * η
+        bb = z - sz / 2 + sz * rand(rng)
+        zz = bb * α
+        finish = 0
```
The compiler is more clever now, but it still might be a good idea to initialize `totaltime` and `startpos` as a float (i.e., 0.0).

A minor stylistic request: change finish to  `finished = false`, and `while finish == 0` to `while !finished`  to make it easier to read

itsdfish commented on src/DDM.jl:
```diff
@@ -125,230 +226,493 @@ function _large_time_pdf(u::T, z::T, K::Int) where {T<:Real}
+            prob = exp(radius * μ / D)
+            prob = prob / (1 + prob)
+            dir_ = 2 * (rand(rng) < prob) - 1
+            l = -1
```
I think `s1`, `s2`, `told`, and `tnew` can also be initialized as 0.0

itsdfish commented on src/DDM.jl:
```diff
@@ -125,230 +226,493 @@ function _large_time_pdf(u::T, z::T, K::Int) where {T<:Real}
+    end
 
-            l = 1 + (s1^(-F)) * tnew 
+    # Initialize output vectors
```
I made a few stylistic changes, and cleaned up some redundant variables. Can you use the follow if it looks OK to you? 


```
function _rand_rejection(rng::AbstractRNG, d::RatcliffDDM; N::Int = 1)
    (;ν, α, τ, z, η, sz, st, σ) = d

    η = η == 0 ? eps() : η

    # Initialize output vectors
    result = zeros(N)
    T = zeros(N)
    XX = fill(0, N)

    # Called sigma in 2001 paper
    D = σ^2 / 2

    # Program specifications
    ϵ = eps()  # precision from 1.0 to next double-precision number

    for n in 1:N
        r1 = randn()
        μ = ν + r1 * η
        bb = z - sz / 2 + sz * rand()
        zz = bb * α
        finished = false
        totaltime = 0.0
        startpos = 0.0
        Aupper = α - zz
        Alower = -zz
        radius = min(abs(Aupper), abs(Alower))

        while !finished
            λ = 0.25 * μ^2 / D + 0.25 * D * π^2 / radius^2
            # eq. formula (13) in 2001 paper with D = sigma^2/2 and radius = Alpha/2
            F = D * π / (radius * μ)
            F = F^2 / (1 + F^2)
            # formula p447 in 2001 paper
            prob = exp(radius * μ / D)
            prob = prob / (1 + prob)
            dir_ = 2 * (rand() < prob) - 1
            l = -1.0
            s2 = 0.0
            s1 = 0.0
            while s2 > l
                s2 = rand()
                s1 = rand()
                tnew = 0.0
                told = 0.0
                uu = 0
                while abs(tnew - told) > ϵ || uu == 0
                    told = tnew
                    uu += 1
                    tnew = told + (2 * uu + 1) * (-1)^uu * s1^(F * (2 * uu + 1)^2)
                    # infinite sum in formula (16) in BRMIC,2001
                end
                l = 1 + s1^(-F) * tnew
            end
            # rest of formula (16)
            t = abs(log(s1)) / λ
            # is the negative of t* in (14) in BRMIC,2001
            totaltime += t
            dir_ = startpos + dir_ * radius
            ndt = τ - st / 2 + st * rand()
            if (dir_ + ϵ) > Aupper
                T[n] = ndt + totaltime
                XX[n] = 1
                finish = true
            elseif (dir_ - ϵ) < Alower
                T[n] = ndt + totaltime
                XX[n] = 2
                finish = true
            else
                startpos = dir_
                radius = minimum(abs.([Aupper, Alower] .- startpos))
            end
        end
    end
    return (choice=XX,rt=T)
end
```

kiante-fernandez replied:
Looks fine to me. I made the corresponding changes.

itsdfish commented:
> @itsdfish I have finally had a chance to take a look through the new codebase; great work on organizing things! Following the conventions, I added the AbstractDDM and moved everything we needed to DDM.jl. I am still failing some of the tests I set up, but I thought I would ping for feedback while I have time to have eyes on this for a few days.

@kiante-fernandez, thanks! I think we are converging on a good system and API. 

I made a few comments throughout. Everything looks good for the most part. I think the primary item is getting the unit tests to pass and adding more if needed.

kiante-fernandez commented:
> > @itsdfish I have finally had a chance to take a look through the new codebase; great work on organizing things! Following the conventions, I added the AbstractDDM and moved everything we needed to DDM.jl. I am still failing some of the tests I set up, but I thought I would ping for feedback while I have time to have eyes on this for a few days.
> 
> @kiante-fernandez, thanks! I think we are converging on a good system and API.
> 
> I made a few comments throughout. Everything looks good for the most part. I think the primary item is getting the unit tests to pass and adding more if needed.

@itsdfish Thank you for the comments. I have been working to resolve some issues, but I have not yet identified what is wrong with the implementation. I thought it might be due to the numerical integration implementation (inspired by [https://github.com/hddm-devs/hddm/blob/master/src/integrate.pxi]), though I have not pinned it down yet. That is just my sense from some of the test results so far.

DominiqueMakowski commented:
Should we aim at merging this before @kiante-fernandez's talk ^^ ?

itsdfish commented:
I'll close this now that it has been moved to #90
